MM-10601: Enter key on hardware android keyboard does not post message #8797

oljw opened the pull request:
#### Summary
- Basically, I've added a 'onKeyUp' method in Android Native side and emitted events when 'enter' key from a hardware keyboard had been pressed.
- Using this event, in the js side, it sends the message when 'enter' event had been received.
- Distinguishing between soft keyboard and hardware keyboard was done by setting the state of keyboard as 'shown' or 'hidden' using React Native's callback methods: 'keyboardDidShow' and 'keyboardDidHide'.

UPDATE: I've added the same functionality for iOS side.
- I added EventEmitModule.m/h classes to emit events to JS side from iOS native.
- From iOS native side, an event is distinguished using 'UIKeyCommand' and emitted to JS when Enter key is pressed using 'NSNotificationCenter'.
- In JS, post_textbox.js class listens to the event from iOS native and handles event as necessary.

#### Ticket Link
[https://github.com/mattermost/mattermost-server/issues/8797]

#### Device Information
This PR was tested on: 
[Galaxy S8 Active(Android 7.0); a bluetooth keyboard],
[Galaxy Note 9(Android 8.1); a bluetooth keyboard(Anker)],
[iPhone X simulator, v12.2],
[iPhone 7 simulator, v12.2]

enahum commented:
Hey @oljw thanks so much for this, we will be reviewing and testing the PR as soon as we can. I just wanted to ask if it is possible to have this same thing done for iOS?

oljw commented:
@enahum sure, I will take a look at the iOS side as well.

oljw commented:
I just finished implementing the iOS part of this issue and I will submit another PR when this current one is finished as it will conflict with each other.
Thank you!

enahum commented:
@oljw is it possible to submit it as prt of the same PR being the same feature.

Thanks

oljw commented:
@enahum will do.
I'll keep you posted.
Thank you!

oljw commented:
@enahum I added new commits to the current PR.
Both Android and iOS versions are implemented.
Thanks!

enahum commented on app/components/post_textbox/post_textbox.js:
```diff
@@ -85,15 +97,23 @@ export default class PostTextbox extends PureComponent {
         if (Platform.OS === 'android') {
-            Keyboard.addListener('keyboardDidHide', this.handleAndroidKeyboard);
             BackHandler.addEventListener('hardwareBackPress', this.handleAndroidBack);
+            this.listenerId = mattermostManaged.addEventListener('hardwareEnter', this.handleHardwareEnterKey);
```
I'm failing to see how this event is emitted by `mattermostManaged`

enahum commented on app/components/post_textbox/post_textbox.js:
```diff
@@ -85,15 +97,23 @@ export default class PostTextbox extends PureComponent {
+        } else if (Platform.OS === 'ios') {
+            const {EventEmitModule} = NativeModules;
+            this.eventEmitter = new NativeEventEmitter(EventEmitModule);
+            this.eventEmitter.addListener('handleIosEnter', this.handleHardwareEnterKey);
```
Do we need a new native module just for this? is there another way to accomplish the same thing? Also try and keep both platforms consistent for example I don't get why the event is called different, this should be `hardwareEnter` the same as Android.

enahum commented on ios/Mattermost/EventEmitModule.h:
```diff
@@ -0,0 +1,14 @@
+#import <React/RCTBridgeModule.h>
+#import <React/RCTEventEmitter.h>
+
+@interface EventEmitModule : RCTEventEmitter <RCTBridgeModule>
```
see my comment previous comment, this can be part of an existing module

oljw commented:
@enahum Thanks for the review, I'll fix them up.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

oljw commented:
Hey @migbot, I am currently studying more into the React Native itself and its communication with native sides at the moment. I will submit the changes soon. Thanks for following up!

oljw commented:
Hi, @enahum, @migbot 

I commited a set of new codes after the first review.
Apologies as it had been awhile since the last commit.

Below are some of my comments about the previous commit and the new commit.

I first thought using the mattermostManaged was a common way throughout the app to communicate with the native side. So I wanted to use the addListener function to communicate with the Android native side.

I have removed the new native module to communicate with ios native side. Instead, I added the functionality inside of the MattermostManaged.m file as its header file extends from RCTEventEmitter which is necessary to communicate with JS side. Additionally, I added a function called addNativeListener in the mattermost-managed.ios.js and .android.js files in JS side as its 'addEventListener' function requires a config value.

I am a little unsure about the fact that I added the 'addNativeListener' function inside of each mattermost-managed files.
However, having this function in the mattermostmanaged makes sure that I don't have to divide the source code depending on each platform.

Please advise me if there are things that needs to be fixed again.

Thank you very much.

oljw commented:
Hi @enahum,
Thanks for the quick review.
I removed the parts where MattermostManaged is used for a communication module between the native and JS.
Instead, I created a new module called 'EventEmitterModule' to register/unregister native event listeners in JS side.
 
About your comment on Apr 7th: 
"Do we need a new native module just for this? is there another way to accomplish the same thing?"

=> As I researched about communication with JS and iOS, this seems to be the best way for now.
And a new native module is necessary for communication because I believe 'MattermostManaged' module is the only one to communicate between them.
So I had to create a new module for communication.

Thanks for your support and let me know if there are room for improvements.

enahum commented on app/components/post_textbox/post_textbox.js:
```diff
@@ -175,8 +185,22 @@ export default class PostTextbox extends PureComponent {
     };
 
-    handleAndroidKeyboard = () => {
-        this.blur();
```
when using the device keyboard instead of a hardware keyboard and then I use the hardware button to hide the keyboard the textbox won't get blurred.

oljw replied:
Hello @enahum,

I ran into an issue when I enabled the blurring when soft-keyboard hides.

Here's the scenario (this only happens in Android):
1. Connect to a bluetooth keyboard.
2. Enable virtual-keyboard to be visible (by tapping keyboard icon bottom-right corner and enable Virtual Keyboard).
3. Tap the TextInput box to focus.
4. Virtual-keyboard will be shown.
5. Type something with hardware keyboard.
6. Virtual-keyboard will auto-hide and TextInput will be blurred.
7. User cannot type anything using the hardware keyboard as this will continue to happen.

The workaround scenario I tried to do was to add a keyboardState state of 'hiding' and check it in onChangeText(as this gets called when input is received before virtual-keyboard is hidden) and handle it in keyboardWillHide. But keyboardWillHide isn't supported on Android platform.

At this point, all I can think of as a solution is to receive every input from Android native side and enable focus when any key is pressed but this means native side will send message every time a key is pressed. This eventually lead to performance issue as it will need to parse every key press.

I'll need some help to resolve this issue.

Thank you in advance!

oljw replied:
Update: I'm not able to reproduce the issue I described above.
I will test more though.

enahum commented on app/components/post_textbox/post_textbox.js:
```diff
@@ -175,8 +185,22 @@ export default class PostTextbox extends PureComponent {
-    handleAndroidKeyboard = () => {
-        this.blur();
+    handleHardwareEnterKey = () => {
+        if (this.state.keyboardState === 'hidden') {
```
we shouldn't care about the state of the software keyboard if I'm using an external keyboard and press Shift+Enter to post the message

enahum replied:
on the other hand there is the need to use the `hidden` state to trigger the below function as this will also apply to pressing the enter key in the soft keyboard

oljw replied:
Yes this is actually necessary just like what you said. When tapping the enter button even on the soft-keyboard, this will get triggered and it won't possible do a line break.

enahum commented on ios/Mattermost/EventEmitterModule.m:
```diff
@@ -0,0 +1,46 @@
+
+- (void)handleHardwareEnter:(NSNotification *)notification
+{
+  [self sendEventWithName:@"hardwareEnter"
```
you should check if there are listeners before emitting the event

oljw replied:
I've added the code to check if there are listeners.

oljw commented:
@migbot Thanks for the comment!

Could you elaborate on that iPad issue?
I'm not sure what the issue is.

On Android part, I've only tested on real devices with bluetooth keyboards but I've never encountered that issue.
However, as I just tested on an Android emulator(Android 8.1, Nexus 5X), I can confirm this issue exists.
As I log the keys in the android native side but it does not detect the shift key correctly.
Also, I tried to find an alternative way to detect the shift key in the emulator, but I couldn't find any solution for now.
If you look at below stackoverflow post, you could see someone having the same issue:
https://stackoverflow.com/q/39491596/5715905
However, even when I test the solution provided in the link, it still cannot detect the shift key alone in the emulator.
When I use the real device, I can log out the shift key press correctly. (I've tested two Android devices with two different bluetooth keyboards)
I will try to look more into this issue.

The newline before submit happens because in Android native part, it's overriding the 'onKeyUp' which gets triggered when the key comes 'up'.
On this issue post: https://github.com/mattermost/mattermost-server/issues/8797#issuecomment-476915781 you've mentioned about overriding more callbacks when necessary.
In order to resolve this issue, it'd help to override 'dispatchKeyEvent' method which can handle both key up and key down.

Thanks again for your comment and let me know if you have any tips!

oljw commented:
@migbot 
Got it!! I'll try to fix that issue.
And I'll look into that reply page issue also.
Thanks!

oljw commented:
@migbot @enahum Here is a summary for the current status:

____resolved issues____
1. The iPad issue with onKeyboardShow had been resolved.
2. The issue that sent the message that isn't in focus had been resolved.

____remaining issues____
1. The part where android sends message on key 'up' cannot be resolved without overriding other native method that can detect key 'down'.

2. I couldn't find a workaround for android emulators not recognizing 'shift+enter'. There is no other way to detect shift key. (This issue doesn't get reproduced with real devices)
- Cannot distinugish using Metastate.
- isShiftPressed() returns false.
- hasNoModifiers() returns true.

3. Issue with blurring on onKeyboardHide:
https://github.com/mattermost/mattermost-mobile/pull/2676#discussion_r283126885
This issue doesn't get reproduced in emulators but it happens in real devices.
In emulators, the software keyboard doesn't hide as soon as the hardware keyboard key is pressed.
I've tried many workaround solutions from giving a delay before blurring to sending every key pressed event from native android side to give focus. But none of these solutions were actually good.

I'd appreciate it if I could get a feedback or an input on these remaining issues.
Thanks.

oljw commented:
Hi @migbot No worries.
I understand your case about creating a separate module just for the text input component.
I will investigate and create a custom component module and see if it'd be acceptable.
I'll let you know if I need any help along the way.
Thanks!

oljw commented:
Hi, @migbot I've been trying to make a custom TextInput component by extending the ReactEditText (Android side). 
I tried various methods and I thought the best way to approach this was to extend the class and make it custom to receive events and such.
However, this approach isn't possible as many classes and methods cannot be accessed from outside of package.
So how should I approach this?

Copy the original classes in React-native repository and customize them as I like in a separate module?
This is all I can think of at the moment but wasn't sure if it is a good approach or not.

Let me know about your thoughts.
Thanks.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

jasonblais commented:
Hey @migbot, wondering if you have thoughts on the above note from @oljw?

oljw commented:
@migbot First of all, I could not find out how to focus the TextInput box. the ReactEditText class only allows the TextInput component to be focused only from the JS side. Which I had to modify private variables in class to bypass that.
Also, there is no prop of getting the 'value' within the JS side.
I believe these needs to be done in index file in React Native side.
I have been searching about how to setup the index file to achieve this, but it is taking a bit too long for me to understand and learn it as I don't have a lot of time due to other responsibilities recently.

So as of now, on my side, I think I need you or someone else to create the module itself and have its basic functionality done and I can add this enter key functionality in that particular module.

Or if you have any other guides that talks about this particular case or similar, I can try to look into it.

oljw commented:
@migbot Great!
I hope you'll make it happen 😄
Let me know if you need any help on anything.
Thanks for the update!

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

jasonblais commented:
Hi @migbot - touching base to see if there are next steps for this PR? Thanks!

enahum commented:
@oljw would you want to continue working on this one or should we re-open the help wanted ticket?

oljw commented:
@enahum I'd say you just re-open a new one. I won't have lots of time due to other responsibilities. Thanks.

enahum commented:
Thanks @oljw
